### PR TITLE
Expose raw data in serial editor

### PR DESCRIPTION
**Pull request by cait1inhennessy**
(1) Graph data like "x:123" now goes to the console too
(2) Export data button on top of console to get raw text 

image

(3) Support for slurping CSV data as well

image

### CODE REVIEW

**Comment by pelikhan** [comment]
I'd like to review this at the design meeting. I was thinking about a button where the default interaction is CSV - with a side dropdown for text.

**Comment by pelikhan** [comment]
Export is an advanced feature. Are we ok with this kind of drop down menu?

File: webapp/src/serial.tsx [hunk 1]
```diff
@@ -136,13 +140,20 @@ export class Editor extends srceditor.Editor {
...
-                return;
             }
         }
 
         this.appendConsoleEntry(data)
     }
 
+    appendRawData(data: string) {
+       this.rawDataBuffer += data;
+       let excessChars = this.rawDataBuffer.length - this.maxBufferLength;
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: trim it by 30% to avoid constant trimming

File: webapp/src/serial.tsx [hunk 2]
```diff
@@ -256,8 +268,8 @@ export class Editor extends srceditor.Editor {
             lines.map(line => `time (s)${sep}${line.name}`).join(sep) + '\r\n';
 
         const datas = lines.map(line => line.line);
-        const nl = datas.map(data => data.length).reduce((l, c) => Math.max(l, c));
-        const nc = this.charts.length;
+        const nl = datas.length > 0 ? datas.map(data => data.length).reduce((l, c) => Math.max(l, c)) : 0;
+        const nc = this.charts.length;"t"
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: random "t"

File: webapp/src/serial.tsx [hunk 3]
```diff
@@ -267,11 +279,20 @@ export class Editor extends srceditor.Editor {
         core.infoNotification(lf("Exporting data...."));
     }
 
+    downloadRaw() {
+        pxt.commands.browserDownloadAsync(this.rawDataBuffer, lf("{id:csvfilename}data") + ".txt", "text/plain");
+        core.infoNotification(lf("Exporting data...."));
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: show before doing download

File: webapp/src/serial.tsx (hunk 1 above)
Comment (cait1inhennessy) [reply webapp/src/serial.tsx]: Thanks, done

File: webapp/src/serial.tsx [hunk 4]
```diff
@@ -148,10 +148,9 @@ export class Editor extends srceditor.Editor {
 
     appendRawData(data: string) {
        this.rawDataBuffer += data;
-       let excessChars = this.rawDataBuffer.length - this.maxBufferLength;
-        if (excessChars > 0) {
-           this.rawDataBuffer = this.rawDataBuffer.slice(excessChars);
-        }
+       if (this.rawDataBuffer.length > this.maxBufferLength) {
+        this.rawDataBuffer.slice(this.rawDataBuffer.length / 4);
```
Comment (pelikhan) [inline webapp/src/serial.tsx]:
```
this.rawDataBuffer = this.rawDataBuffer.slice(this.rawDataBuffer.length / 4);
```

**Comment by pelikhan** [comment]
@abchatra @samelhusseini see new button on top of the console to allow downloading the raw output. We should really have a way to drag up and down the console but that can be for another PR.

**Comment by abchatra** [comment]
Lets take this to design review today. Is this bumped in beta to visualize?

**Comment by pelikhan** [comment]
Yes

**Pull request merged**